### PR TITLE
Fixed code for multinode

### PR DESCRIPTION
**Pull request by marcoSanti**
This commit introduces the following changes:

- More logging has been added
- In open.hpp:update_file_metadata() a branch had  writers[pid][path] set to false on both branches of if.  this  was wrong and caused the exit handler to not properly set a file to be commited (completed) once the program normally terminated.
- in server/comm/remote_listener.hpp::remote_listener_remote_sending() function, file_shm allways resulted in being not initialized. which caused an error in mpi recive. This was due to missing else branch on checking for file size.

### CODE REVIEW

File: src/server/comm/backends/mpi.hpp [hunk 1]
```diff
@@ -96,15 +98,22 @@ class MPI_backend : public backend_interface {
...
         for (long int k = 0; k < nbytes; k += elem_to_snd) {
-            (nbytes - k > MPI_MAX_ELEM_COUNT) ? elem_to_snd = MPI_MAX_ELEM_COUNT
-                                              : elem_to_snd = nbytes - k;
 
+            if (nbytes - k > MPI_MAX_ELEM_COUNT) {
+                elem_to_snd = static_cast<int>(MPI_MAX_ELEM_COUNT);
+            } else {
+                elem_to_snd = static_cast<int>(nbytes - k);
+            }
+
```
Comment (GlassOfWhiskey) [inline src/server/comm/backends/mpi.hpp:112]: The block above can be replaced with

```c++
elem_to_snd = static_cast<int>(std::min(nbytes - k, MPI_MAX_ELEM_COUNT));
```

File: src/server/comm/backends/mpi.hpp [hunk 2]
```diff
@@ -255,22 +267,34 @@ class MPI_backend : public backend_interface {
...
         for (long int k = 0; k < bytes_expected; k += bytes_received) {
-            (bytes_expected - k > MPI_MAX_ELEM_COUNT) ? count = MPI_MAX_ELEM_COUNT
-                                                      : count = bytes_expected - k;
 
+            if (bytes_expected - k > MPI_MAX_ELEM_COUNT) {
+                count = static_cast<int>(MPI_MAX_ELEM_COUNT);
+            } else {
+                count = static_cast<int>(bytes_expected - k);
+            }
+
```
Comment (GlassOfWhiskey) [inline src/server/comm/backends/mpi.hpp:291]: Same here:

```c++
count = static_cast<int>(std::min(bytes_expected - k, MPI_MAX_ELEM_COUNT));
```

File: src/common/capio/logger.hpp [hunk 3]
```diff
@@ -17,7 +17,7 @@
 #endif
 
 #ifndef __CAPIO_POSIX
-std::ofstream logfile; // if building for server, self contained logfile
+static thread_local std::ofstream logfile; // if building for server, self contained logfile
```
Comment (GlassOfWhiskey) [inline src/common/capio/logger.hpp:20]: Please move all updates to `logger.hpp` in #33

File: src/server/capio_server.cpp [hunk 4]
```diff
@@ -193,7 +193,8 @@ int parseCLI(int argc, char **argv, int rank) {
         auto hostname = new char[HOST_NAME_MAX];
         gethostname(hostname, HOST_NAME_MAX);
 
-        std::string filename = token + "_" + hostname + ".log";
+        std::string filename =
```
Comment (GlassOfWhiskey) [inline src/server/capio_server.cpp:196]: Move also these updates to #33 
PS: can't `filename` be declared `const`?

**Review by GlassOfWhiskey — CHANGES_REQUESTED** [review]

File: src/common/capio/logger.hpp [hunk 5]
```diff
@@ -83,11 +82,9 @@ class Logger {
         if (!logfile.is_open()) {
             // NOTE: should never get to this point as capio_server opens up the log file while
             // parsing command line arguments. This is only for failsafe purpose
-            auto hostname = new char[HOST_NAME_MAX];
-            gethostname(hostname, HOST_NAME_MAX);
-            logfile.open(std::string(CAPIO_LOG_SERVER_DEFAULT_FILE_NAME) + hostname + ".log",
+            logfile.open(std::string(CAPIO_LOG_SERVER_DEFAULT_FILE_NAME) + std::to_string(tid) +
```
Comment (GlassOfWhiskey) [inline src/common/capio/logger.hpp:85]: Here you are already passing the `tid` to the filename.
Please move this logic to #33 which is the correct place to discuss it

File: src/server/capio_server.cpp [hunk 6]
```diff
@@ -207,16 +208,12 @@ int parseCLI(int argc, char **argv, int rank) {
     } else {
 #ifdef CAPIOLOG
         // log file not given. starting with default name
-        auto hostname = new char[HOST_NAME_MAX];
-        gethostname(hostname, HOST_NAME_MAX);
-
         const std::string logname =
-            CAPIO_LOG_SERVER_DEFAULT_FILE_NAME + std::string(hostname) + ".log";
+            CAPIO_LOG_SERVER_DEFAULT_FILE_NAME + std::to_string(capio_syscall(SYS_gettid)) + ".log";
```
Comment (GlassOfWhiskey) [inline src/server/capio_server.cpp:212]: Here you are already passing the `tid` to the filename.
Please move this logic to #33 which is the correct place to discuss it

File: src/server/utils/capio_file.hpp [hunk 7]
```diff
@@ -92,6 +93,21 @@ class Capio_file {
         }
     }
 
+    inline void set_complete() {
```
Comment (GlassOfWhiskey) [inline src/server/utils/capio_file.hpp:96]: Please remove this method and make `_complete` default to `true`

```c++
inline void set_complete(bool _complete = true) {
   ...
}
```

**Review by GlassOfWhiskey — CHANGES_REQUESTED** [review]

File: src/server/utils/capio_file.hpp (hunk 7 above)
Comment (marcoSanti) [reply src/server/utils/capio_file.hpp:96]: This should not be done, as if a capio file is set to true by default, then a file that has just been created but not yet written will be available to other applications, completely ignoring the semantic specified on the file. The correct behavior is the one implemented.

**Review by GlassOfWhiskey — APPROVED** [review]

**Pull request merged**